Models/Assimilation/Assimilation.cs

```diff
@@ -0,0 +1,109 @@
+
+        public bool TryUpdateAssimilation(ILeafWaterInteraction Water, PhotosynthesisParams Params)
+        {
+            AssimilationCalculator calc;
+
+            // TODO: There might be a better structure for this, since it's only used to
+            // determine the assimilation parameters
+            if (Canopy.Type == CanopyType.C3) 
+                calc = new CalculatorC3(this, Partial);
+            else if (Canopy.Type == CanopyType.C4) 
+                calc = new CalculatorC4(this, Partial);
+            else 
+                calc = new CalculatorCCM(this, Partial);
+
+            var aparam = calc.GetAssimilationParams();
+
+            double resistance;
+
+            // If there is no limit on the water supply
+            if (!Params.limited)
+            {
+                IntercellularCO2 = Path.IntercellularToAirCO2Ratio * Canopy.AirCO2;
+
+                aparam.p = IntercellularCO2;
+                aparam.q = 1 / MesophyllCO2ConductanceAtT;
+
+                CO2AssimilationRate = aparam.CalculateAssimilation();
+                resistance = Water.UnlimitedWaterResistance(CO2AssimilationRate, Canopy.AirCO2, IntercellularCO2);
+                WaterUse = Water.HourlyWaterUse(resistance, Partial.AbsorbedRadiation);
+            }
+            // If water supply is limited
+            else
+            {
+                WaterUse = Params.maxHourlyT * Params.fraction;
+                var WaterUseMolsSecond = WaterUse / 18 * 1000 / 3600;
+
+                resistance = Water.LimitedWaterResistance(WaterUse, Partial.AbsorbedRadiation);
+                var Gt = Water.TotalLeafCO2Conductance(resistance);
+
+                aparam.p = Canopy.AirCO2 - WaterUseMolsSecond * Canopy.AirCO2 / (Gt + WaterUseMolsSecond / 2.0);
+                aparam.q = 1 / (Gt + WaterUseMolsSecond / 2) + 1.0 / MesophyllCO2ConductanceAtT;
+
+                CO2AssimilationRate = aparam.CalculateAssimilation();
+
+                if (!(calc is CalculatorC3))
+                    IntercellularCO2 = ((Gt - WaterUseMolsSecond / 2.0) * Canopy.AirCO2 - CO2AssimilationRate) / (Gt + WaterUseMolsSecond / 2.0);
+            }
+
+            // C4 & CCM
+            if (!(calc is CalculatorC3))
+                MesophyllCO2 = IntercellularCO2 - CO2AssimilationRate / MesophyllCO2ConductanceAtT;
+
+            // CCM ONLY
+            if (calc is CalculatorCCM)
+            {
+                ChloroplasticO2 = Path.PS2ActivityInBundleSheathFraction * CO2AssimilationRate / (Canopy.DiffusivitySolubilityRatio * calc.Gbs) + Canopy.OxygenPartialPressure;
+                ChloroplasticCO2 = MesophyllCO2 + (MesophyllCO2 * aparam.x4 + aparam.x5 - aparam.x6 * CO2AssimilationRate - aparam.m - aparam.x7) * aparam.x8 / calc.Gbs;
+            }
+
+            // New leaf temperature
+            LeafTemperature = (Water.LeafTemperature(resistance, Partial.AbsorbedRadiation) + LeafTemperature) / 2.0;
+
+            // If the assimilation is not sensible
```

Comment:
Should this raise an exception?

Reply:
Don't think so. Alex explicitly said nonsensical results were a possible result of the model, and that in such cases the values should simply be set to 0, which is supposed to be doing. It's returning true/false rather than directly setting to 0 in the code because the results of multiple assimilations need to be tested. 

In hindsight it could just set the values to 0 and then test for 0's, rather than a true or false result, which might make more sense to read. The current pattern is a legacy of the old version.

Reply:
I really don't like that it can return NANs or -ve's - that just feels wrong.
If that is what Alex says, then the true/false is fine and probably more clear on what is happening.